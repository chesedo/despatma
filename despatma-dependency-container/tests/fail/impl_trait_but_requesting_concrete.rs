struct Config {
    port: u32,
}

trait DAL {}

struct PostgresDAL;

impl DAL for PostgresDAL {}

struct Service<D: DAL> {
    dal: D,
}

impl<D: DAL> Service<D> {
    fn new(port: u32, dal: D) -> Self {
        println!("Impl trait service started on port {}", port);
        Self { dal }
    }
}

#[despatma_dependency_container::dependency_container]
impl DependencyContainer {
    fn config(&self) -> Config {
        Config { port: 8080 }
    }

    #[Transient(PostgresDAL)]
    fn dal(&self) -> impl DAL {
        PostgresDAL
    }

    fn service(&self, config: Config, dal: PostgresDAL) -> Service<PostgresDAL> {
        Service::new(config.port, dal)
    }
}

fn main() {
    let container = DependencyContainer::new();
    let _service = container.service();
}
