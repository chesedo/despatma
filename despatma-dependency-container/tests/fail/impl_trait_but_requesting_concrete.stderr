error: Requested type is a concrete type, but the registered type is: `impl DAL`

         = help: change this to `dal: impl DAL`

  --> tests/fail/impl_trait_but_requesting_concrete.rs:33:39
   |
33 |     fn service(&self, config: Config, dal: PostgresDAL) -> Service<PostgresDAL> {
   |                                       ^^^^^^^^^^^^^^^^

error[E0433]: failed to resolve: use of undeclared type `DependencyContainer`
  --> tests/fail/impl_trait_but_requesting_concrete.rs:39:21
   |
39 |     let container = DependencyContainer::new();
   |                     ^^^^^^^^^^^^^^^^^^^ use of undeclared type `DependencyContainer`
